crates/adapters/src/controller/error.rs

```diff
@@ -117,8 +156,22 @@ impl ConfigError {
 }
 
 /// Controller error.
-#[derive(Debug)]
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
 pub enum ControllerError {
+    /// I/O error.
+    IoError {
+        /// Describes the context where the error occurred.
+        context: String,
```

Comment:
from the thiserror example; the context is just a backtrace which seems better as it's automatic and more informative
```
    Io {
        #[from]
        source: io::Error,
        backtrace: Backtrace,
    },
```

Reply:
So this won't quite work, because I want all errors to implement `Serialize`, which is tricky for `io::Error` (requires implementing `Serialize`) manually.

With `context` I was trying to make errors human-readable.  But maybe `Backtrace` is better.